docs/DesignDocs/NotesOnAttestationAPI.md

```diff
@@ -0,0 +1,160 @@
+configured for specified TEEs and platforms, and to generate or verify
+the evidence formats according to their defined security policies.
+  - The OE SDK attestation API enables the attesters and verifiers to discover
+  and register supported plugins that are built, configured, and working
+  properly. For a specific attester or verifier instance, the supported list
+  of evidence is not required to perfectly match the list defined in the
+  security policies.
+
+# Options for Enclave-side Plugin Library Initialization
+
+As described in the design document
+[Attestation: OE SDK Integration with Intel® SGX SDK quote-ex Library for Generation of Evidence in New Formats](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/SGX_QuoteEx_Integration.md),
+enclave-side plugin library initialization of attester plugins can be
+triggered by application enclave call of the OE SDK API
+`oe_get_attester_plugins()`.
+During initialization, the plugin library enumerates supported evidence
+formats, and returns a list of supported attester plugins.
+
+In general, there are several options for initialization of enclave-side
+plugin libraries while keeping application enclaves TEE agnostic.
+
+In this section, we focus on the discussion of the options for initialization
+of attester plugins. Similar options are available for initialization of
+verifier plugins.
+
+1. Every enclave-side plugin library implements the same OE SDK API
+`oe_get_attester_plugins()`.
+
+    With this option, a build of an application enclave project can only link
+    with a single attestation plugin library, to avoid function name collision.
+
+2. Compile-time macro to convert from TEE-agnostic API to TEE-specific
+implementation.
+
+    With this option, application enclaves still call TEE-agnostic OE SDK API
+    `oe_get_attester_plugins()`.
+    But the OE SDK provides a definition of this function as a macro calling
+    into the TEE- and library-specific initialization functions of linked
+    enclave-side plugin libraries.
+
+    This option has the potential to support linking with multiple
+    enclave-side plugin libraries in a single application enclave build
+    configuration.
+    The OE SDK build environment is required to provide a proper
+    `oe_get_attester_plugins()` macro definition for every configuration
+    it supports.
+
+3. Every enclave-side plugin library registers its initialization function
+with the OE SDK framework.
+
+    The design document
+    [OpenEnclave Modules](https://github.com/jhand2/openenclave/blob/global_module_initialization/docs/DesignDocs/openenclave_modules.md)
+    describes how an enclave-side module can register its initialization
+    function for invocation on first entry to the enclave (upon enclave load).
+
+    With this option, the implementation of the OE SDK attestation framework
+    provides a hook for enclave-side plugin libraries to register their
```

Comment:
>enclave-side plugin libraries [](start = 24, length = 29)

The same initialization should also be supported on the host-side plugin libraries. We have to recommend a universal way (for all platforms) for libraries to initialize on load. Typical ways to do initialization on load are 1) global variables 2) .init_array. The order of such initialization is critical. If the plugin libraries try to register before OE SDK is ready, it's a potential issue.

I'd suggest OE SDK team to finalize the design in [https://github.com/jhand2/openenclave/blob/global_module_initialization/docs/DesignDocs/openenclave_modules.md](https://github.com/jhand2/openenclave/blob/global_module_initialization/docs/DesignDocs/openenclave_modules.md) and prototype this on Linux and Windows to ensure this is a viable option.

Reply:
I have the global module initialization work planned for v0.10 release. I will include you and @shnwc in the reviews.

Note that my current design is only for enclave-side initialization. Host-side initialization for Windows is more complicated.

Reply:
@yentsanglee agreed that host-side support is needed. Will edit the doc to reflect that. 

@jhand2 what is the plan to address the Linux and Windows host-side initialization?

Reply:
@yentsanglee Can the function `void register_plugin_function(func)` be implemented to initialize its `_lock` and `_registered_function_list` upon the first call by an plugin library?

Reply:
If register_plugin_function is called in reaction to another global variable initialization, I am not sure how to universally guarantee _lock and _registered_function_list are initialized. Any suggestion?

Reply:
One example implementation would be for the metadata (.conf file) to contain the list of modules at sign time.   Then global initialization code enumerates the list from the metadata, and invokes the initialization functions in the order of that list.   Then the app writer does not need to be involved (only the signer), and you get ordering if it is needed.

Reply:
From what I understand, the developers have to:
1. Create a .conf file in a pre-defined syntax.
2. Links to the compiled metadata (.conf file)

OE SDK has to:
1. Have a signtool that compiles the .conf file
2. Invoke initialization functions.

Plugin authors have to:
1. Export an initialization function and avoid naming conflict

I think this deserves a separate design. This sounds like a plugin model for generic purpose (not just attestation).

Reply:
@jhand2 how about including this proposed approach in your [OpenEnclave Modules](https://github.com/jhand2/openenclave/blob/global_module_initialization/docs/DesignDocs/openenclave_modules.md) document?

Reply:
@yentsanglee upon enclave load, how does the OE SDK framework map an initialization function from its name listed in the .conf file to the entry point address  for invocation?

Reply:
The main barrier in my original design for modules what that there was no way to tell the linker to keep the initialization functions since we link with `--gc-sections` and there is no reference to the function (it's called by address). I see the same barrier in this design on the enclave side.

Also does this proposal have a host-side solution?

I also have the same question as @shnwc and I don't have a great answer right now.

Reply:
@jhand2 Just committed a new version that contains host-side options.

Reply:
I am not sure if .conf or metadata would improve the development experience but it has to work for enclave and host, Windows and Linux.

I suggest we redesign a plugin generically for OE SDK while the additional attestation formats continue to develop without introducing (much) new public APIs.

Reply:
@yentsanglee before option 3 support is ready for both enclave and host sides, we can readily support option 1, with the limitation that only one plugin library can be linked (which is currently the case for SGX). I'll update the document to capture that.

Can you elaborate on  second paragraph "I suggest we redesign a plugin ..."? Since the plugin APIs are all experimental, I'd like to see a new version of the API definition published soon, so the  SGX plugins implementation in my fork can be upstreamed.

Reply:
Since there are multiple instances of plugin design in OE SDK (for different TEEs, logging, etc), I think we should start a new overall design to set the plugin story straight for all purposes. For the EPID and Quote-ex work, I suggest we continue the work in your fork and integrate with minimum API changes.

I am fine with option 1 but that's where the disagreement started.

Reply:
Let's start with option 1 as experimental in short-term, while working out a longer-term option.